File: app01_java_nft/backend/src/main/java/com/gandarych/nftcerts/watermark/MetadataWatermarkService.java
package com.gandarych.nftcerts.watermark;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.imaging.Imaging;
import org.apache.commons.imaging.common.ImageMetadata;
import org.apache.commons.imaging.formats.jpeg.JpegImageMetadata;
import org.apache.commons.imaging.formats.jpeg.exif.ExifRewriter;
import org.apache.commons.imaging.formats.tiff.TiffImageMetadata;
import org.apache.commons.imaging.formats.tiff.constants.TiffTagConstants;
import org.apache.commons.imaging.formats.tiff.write.TiffOutputDirectory;
import org.apache.commons.imaging.formats.tiff.write.TiffOutputSet;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Embeds a small JSON payload linking a photograph to its content hash and the artist's DID
 * directly into the file's EXIF metadata, providing an additional (non-cryptographic) provenance
 * signal alongside the on-chain/IPFS record.
 *
 * <p>Implementation note: Apache Commons Imaging exposes {@code EXIF_TAG_USER_COMMENT} as a
 * {@code TagInfoGpsText} field, which requires the EXIF "character code" prefix framing rather
 * than a plain string. To keep this straightforward and robust across the resolved library
 * version (1.0.0-alpha5), this implementation instead writes the JSON payload into the root
 * IFD0 {@code ImageDescription} tag ({@link TiffTagConstants#TIFF_TAG_IMAGE_DESCRIPTION}), which
 * is a plain ASCII field and is well supported for lossless rewriting via {@link ExifRewriter}.
 */
@Service
public class MetadataWatermarkService {

    private static final Logger log = LoggerFactory.getLogger(MetadataWatermarkService.class);
    private static final String JPEG_CONTENT_TYPE = "image/jpeg";

    private final ObjectMapper objectMapper;

    public MetadataWatermarkService(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    /**
     * Embeds {@code {"contentHash":"<hex>","artistDid":"<did>"}} into the EXIF ImageDescription
     * field of a JPEG. For any other content type, the original bytes are returned unchanged and
     * a warning is logged (documented behavior, not a bug).
     */
    public byte[] watermark(byte[] originalContent, String contentType, String contentHash, String artistDid) {
        if (!JPEG_CONTENT_TYPE.equalsIgnoreCase(contentType)) {
            log.warn("Skipping EXIF watermark embedding for unsupported content type '{}': only {} is supported",
                    contentType, JPEG_CONTENT_TYPE);
            return originalContent;
        }

        try {
            String payloadJson = buildPayloadJson(contentHash, artistDid);

            TiffOutputSet outputSet = resolveOutputSet(originalContent);
            TiffOutputDirectory rootDirectory = outputSet.getOrCreateRootDirectory();
            rootDirectory.removeField(TiffTagConstants.TIFF_TAG_IMAGE_DESCRIPTION);
            rootDirectory.add(TiffTagConstants.TIFF_TAG_IMAGE_DESCRIPTION, payloadJson);

            ByteArrayOutputStream output = new ByteArrayOutputStream();
            new ExifRewriter().updateExifMetadataLossless(originalContent, output, outputSet);
            return output.toByteArray();
        } catch (IOException e) {
            log.warn("Failed to embed EXIF watermark, returning original file content unchanged", e);
            return originalContent;
        }
    }

    private String buildPayloadJson(String contentHash, String artistDid) {
        Map<String, String> payload = new LinkedHashMap<>();
        payload.put("contentHash", contentHash);
        payload.put("artistDid", artistDid);
        try {
            return objectMapper.writeValueAsString(payload);
        } catch (IOException e) {
            throw new IllegalStateException("Failed to serialize watermark payload", e);
        }
    }

    private TiffOutputSet resolveOutputSet(byte[] originalContent) throws IOException {
        ImageMetadata metadata = Imaging.getMetadata(originalContent);
        if (metadata instanceof JpegImageMetadata jpegImageMetadata) {
            TiffImageMetadata exif = jpegImageMetadata.getExif();
            if (exif != null) {
                TiffOutputSet existing = exif.getOutputSet();
                if (existing != null) {
                    return existing;
                }
            }
        }
        return new TiffOutputSet();
    }
}

File: app03_csharp_nft/backend/NftCerts/Watermark/MetadataWatermarkService.cs
using System.Text.Json;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Formats.Jpeg;
using SixLabors.ImageSharp.Metadata.Profiles.Exif;

namespace NftCerts.Watermark;

// Embeds a JSON payload referencing the content hash and artist DID into the JPEG's
// Exif.Image.ImageDescription tag via ImageSharp. Metadata-based (not steganographic) by design —
// see the ADR in docs/sdlc/10-glossary-and-decisions.md. Unsupported content types and unreadable
// images gracefully pass the original bytes through, matching app01/app02 behavior.
public class MetadataWatermarkService(ILogger<MetadataWatermarkService> logger)
{
    private const string JpegContentType = "image/jpeg";

    public byte[] Watermark(byte[] originalContent, string contentType, string contentHash, string artistDid)
    {
        if (!string.Equals(contentType, JpegContentType, StringComparison.OrdinalIgnoreCase))
        {
            logger.LogInformation(
                "Skipping EXIF watermark embedding for unsupported content type '{ContentType}': only {Supported} is supported",
                contentType, JpegContentType);
            return originalContent;
        }

        try
        {
            string payloadJson = JsonSerializer.Serialize(new { contentHash, artistDid });

            using var image = Image.Load(originalContent);
            image.Metadata.ExifProfile ??= new ExifProfile();
            image.Metadata.ExifProfile.SetValue(ExifTag.ImageDescription, payloadJson);

            using var output = new MemoryStream();
            image.Save(output, new JpegEncoder());
            return output.ToArray();
        }
        catch (Exception exception) when (exception is not OutOfMemoryException)
        {
            logger.LogWarning(exception,
                "Failed to embed EXIF watermark, returning original file content unchanged");
            return originalContent;
        }
    }
}
